etf-sdk/src/lib.rs

```diff
@@ -1,237 +1,271 @@
-//     //             let ct_bytes = serde_wasm_bindgen::to_value(&t.aes_ct.ciphertext).unwrap();
-//     //             let nonce_bytes = serde_wasm_bindgen::to_value(&t.aes_ct.nonce).unwrap();
-//     //             let capsule_bytes = serde_wasm_bindgen::to_value(&t.etf_ct).unwrap();
-//     //             // calc valid secrets d = sQ
-//     //             let secrets: Vec<Vec<u8>> = slot_ids.iter().map(|id| {
-//     //                 let q = hash_to_g1(&id);
-//     //                 let d = q.mul(s);
-//     //                 convert_to_bytes::<G1, 48>(d.into()).to_vec()
-//     //             }).collect::<Vec<_>>();
-
-//     //             let sks = serde_wasm_bindgen::to_value(&secrets).unwrap();
-
-//     //             match etf.decrypt(ct_bytes, nonce_bytes, capsule_bytes, sks) {
-//     //                 Ok(m_js) => {
-//     //                     let m: Vec<u8> = serde_wasm_bindgen::from_value(m_js).unwrap();
-//     //                     assert_eq!(m, b"test".to_vec());
-//     //                 }, 
-//     //                 Err(_) => {
-//     //                     panic!("test should pass, but decryption failed");
-//     //                 }
-//     //             }
-//     //         },
-//     //         Err(_) => {
-//     //             panic!("test should pass, but encryption failed");
-//     //         }
-//     //     }
-//     // }
-// }
\ No newline at end of file
+use ark_serialize::CanonicalSerialize;
+use serde::Deserialize;
+#[cfg_attr(tarpaulin, skip)]
+use wasm_bindgen::prelude::*;
+use etf_crypto_primitives::{self, encryption::tlock::{DecryptionResult, SecretKey, TLECiphertext}, ibe::fullident::{IBESecret, Identity}, utils::{self, *}};
+use w3f_bls::{EngineBLS, TinyBLS377};
+use rand_chacha::ChaCha20Rng;
+use ark_std::rand::SeedableRng;
+use w3f_bls::{DoublePublicKey, DoublePublicKeyScheme};
+use serde::Serialize;
+use serde_big_array::BigArray;
+use ark_serialize::CanonicalDeserialize;
+
+/// The encrypt wrapper used by the WASM blob to call tlock.rs encrypt function in etf-crypto-primitives
+/// * 'id_js': ID string for which the message will be encrypted
+/// * 'message_js': Message which will be encrypted
+/// * 'sk_js': secret key passed in from UI. This should be obtained elsewhere later on.
+/// * 'p_pub_js': the public key commitment for the IBE system
+#[wasm_bindgen]
+pub fn encrypt( 
+    id_js: JsValue,
+    message_js: JsValue, // &[u8]
+    sk_js: JsValue,
+    p_pub_js: JsValue
+) -> Result<JsValue, JsError> {
+    // msk => master secret key
+    let msk_bytes: [u8;32] = serde_wasm_bindgen::from_value(sk_js.clone())
+        .map_err(|_| JsError::new("could not decode secret key"))?;
+    let mut rng: ChaCha20Rng = ChaCha20Rng::from_seed(msk_bytes);
+    let msk = convert_from_bytes::<<TinyBLS377 as EngineBLS>::Scalar,32>(&msk_bytes.clone()).ok_or(JsError::new("Could not convert secret key"))?;
+    let secret_key = SecretKey::<TinyBLS377>(msk);
+    // Look into String Formatting
```

Comment:
do we still need to do this?

Reply:
This is no longer necessary since each failure has a unique error message. However if we would want to have more granular and detailed errors, we could revisit this. The comment has been removed from the code.